### PR TITLE
refactor UI markup and add support for BIP21

### PR DESCRIPTION
**Pull request by talvasconcelos**
...

Related to lnbits/lnbits#1812

### CODE REVIEW

File: templates/satspay/display.html [hunk 1]
```diff
@@ -104,9 +175,9 @@
           </div>
           <q-separator></q-separator>
         </div>
-      </div>
+      </div> -->
     </q-card>
-    <q-card class="q-pa-lg" v-if="lnbtc">
+    <!-- <q-card class="q-pa-lg" v-if="lnbtc">
```
Comment (motorina0) [inline templates/satspay/display.html]: remove dead code

File: templates/satspay/display.html [hunk 2]
```diff
@@ -249,6 +371,12 @@
           hostname += '/testnet'
         }
         return hostname
+      },
+      unifiedQR() {
+        let bitcoin = this.charge.onchainaddress && this.charge.onchainaddress.toUpperCase()
```
Comment (motorina0) [inline templates/satspay/display.html]:
```suggestion
const bitcoin = (this.charge.onchainaddress || '').toUpperCase()
```

File: templates/satspay/display.html [hunk 3]
```diff
@@ -249,6 +371,12 @@
...
+        let lightning = this.charge.payment_request && this.charge.payment_request.toUpperCase()
+        let amount = this.charge.amount / 1e8
+        return `bitcoin:${bitcoin}?amount=${amount}&lightning=${lightning}`
```
Comment (motorina0) [inline templates/satspay/display.html]: - what does it show if `${bitcoin}` or `${lightning}` are undefined?
   - is that in accordance with the BIP21 spec?

File: templates/satspay/display.html [hunk 4]
```diff
@@ -221,6 +292,56 @@
 <script>
   Vue.component(VueQrcode.name, VueQrcode)
 
+  Vue.component('satspay-paid',
```
Comment (motorina0) [inline templates/satspay/display.html]: why not have the components in their own files?

Comment (talvasconcelos) [reply templates/satspay/display.html]: I'm just now learning how to do components... may do that further ahead!

File: templates/satspay/display.html (hunk 3 above)
Comment (talvasconcelos) [reply templates/satspay/display.html]: according to the spec (https://bitcoinqr.dev/) it should show just the bitcoin address... i'm still figuring it out... focusing on getting the markup nicer.

**Comment by talvasconcelos** [comment]
Closes #25

**Comment by talvasconcelos** [comment]
Card now looks like this, with Unified QR code if onchain is used:

uqr

Time elapsed:
time

Paid:
paid

**Comment by motorina0** [comment]
- `mempool.js` was added in this PR https://github.com/lnbits/satspay/pull/26
- please remove it from this PR, it will make it easier to review

File: migrations.py [hunk 5]
```diff
@@ -93,3 +93,22 @@ async def m006_add_zeroconf_column(db):
         )
     except OperationalError:
         pass
+
+
+async def m007_add_pending_column(db):
```
Comment (motorina0) [inline migrations.py]: - if I understand correctly this change is unrelated to the this PR
   - it was supposed to be here: https://github.com/lnbits/satspay/pull/22/files#diff-d7af58842a76f9c7d30d94d9e2539190b7f4d85a238ef85dc70e4dad3b026884
- we should have a release with this change alone ASAP (since it is a blocker bug)

File: migrations.py [hunk 6]
```diff
@@ -93,3 +93,22 @@ async def m006_add_zeroconf_column(db):
...
+    """
+    Add 'pending' column for storing the pending amount
+    """
+    try:
+        await db.execute(
+            f"ALTER TABLE satspay.charges ADD COLUMN pending {db.big_int} NOT NULL DEFAULT 0;"
+        )
+
+        await db.execute(
```
Comment (motorina0) [inline migrations.py]: - I think this is redundant,  the value should be `0` by default

File: migrations.py (hunk 5 above)
Comment (motorina0) [reply migrations.py]: hmm, I can see this is on main already

File: migrations.py (hunk 6 above)
Comment (talvasconcelos) [reply migrations.py]: just the `ALTER TABLE...` then , right?

File: static/js/components.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            </span>
+            <span>{{charge.timeLeft}}</span>
+          </div>
+          
+          </div>
+      </q-linear-progress>
+    </div>
+    `,
+  created() {
+    console.log(this.charge.time_left)
```
Comment (motorina0) [inline static/js/components.js]: - remove line

File: templates/satspay/display.html [hunk 8]
```diff
@@ -1,360 +1,234 @@
...
+{% endif %}
 <style>
   header button.q-btn-dropdown {
     display: none;
   }
 </style>
 {% endblock %} {% block scripts %}
 
-<script src="{{ url_for('satspay_static', path='js/mempool.js') }}"></script>
+<script src="https://mempool.space/mempool.js"></script>
```
Comment (motorina0) [inline templates/satspay/display.html]: - merge error
```suggestion
<script src="{{ url_for('satspay_static', path='js/mempool.js') }}"></script>
```

File: static/js/components.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
+Vue.component(VueQrcode.name, VueQrcode)
+
+Vue.component('satspay-paid', {
+  props: ['charge', 'timer'],
+  template: `
+    <div class="col-sm-10 col-lg-8">
```
Comment (motorina0) [inline static/js/components.js]: - the component should not force any col size
- having a `col` at top level also means the parent component has to have a `row`
```suggestion
    <div>
```

File: templates/satspay/display.html [hunk 10]
```diff
@@ -1,360 +1,234 @@
...
+          <q-tab-panel name="uqr">
+            <div class="row justify-center q-mt-sm">
+              <div v-if="!charge.timeLeft && !charge.paid">
+                <q-icon
+                  name="block"
+                  style="color: #ccc; font-size: 21.4em"
+                ></q-icon>
+              </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
```
Comment (motorina0) [inline templates/satspay/display.html]: - move the column layout outside the `satspay-paid` component
```html
<div class="col-sm-10 col-lg-8"...
```

File: templates/satspay/display.html [hunk 11]
```diff
@@ -1,360 +1,234 @@
...
-            </div>
-            <div v-else>
-              <div class="row text-center q-mb-sm">
-                <div class="col text-center">
-                  <span class="text-subtitle2"
-                    >Pay this lightning-network invoice:</span
-                  >
-                </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
```
Comment (motorina0) [inline templates/satspay/display.html]: - move the column layout outside the `satspay-paid` component
```html
<div class="col-sm-10 col-lg-8"...
```

File: templates/satspay/display.html [hunk 12]
```diff
@@ -1,360 +1,234 @@
...
-                <div class="col text-center">
-                  <q-btn
-                    outline
-                    color="grey"
-                    @click="copyText(charge.onchainaddress)"
-                    >Copy address</q-btn
-                  >
-                </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
```
Comment (motorina0) [inline templates/satspay/display.html]: - move the column layout outside the `satspay-paid` component
```html
<div class="col-sm-10 col-lg-8"...
```

File: static/js/components.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        </div>
+        </div>
+    </div>
+    `
+})
+
+Vue.component('satspay-show-qr', {
+  props: ['charge', 'tab', 'value', 'href'],
+  template: `
+    <div class="col-sm-10 col-lg-8">
```
Comment (motorina0) [inline static/js/components.js]: - see the comment on the root element of `'satspay-paid'`

File: static/js/components.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
+Vue.component(VueQrcode.name, VueQrcode)
+
+Vue.component('satspay-paid', {
+  props: ['charge', 'timer'],
+  template: `
+    <div class="col-sm-10 col-lg-8">
+        <q-icon name="check" style="color: green; font-size: 21.4em" class="fit"></q-icon>
+        <div class="row text-center q-mt-lg">
+        <div class="col text-center">
```
Comment (motorina0) [inline static/js/components.js]: - this `<div>` should be indented one level

File: static/js/components.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+            <q-btn outline v-if="charge.completelink" :loading="charge.paid" :percentage="timer" type="a"
```
Comment (motorina0) [inline static/js/components.js]: - previously it was:
```html
<q-btn outline v-if="charge.webhook" 
```
- is the change on purpose (looks right)?

File: static/js/components.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+            </template>{%endraw%}</q-btn>
+            <p v-if="charge.completelink" class="q-pt-md">Redirecting after 5 seconds</p>
+        </div>
+        </div>
+    </div>
+    `
+})
+
+Vue.component('satspay-show-qr', {
+  props: ['charge', 'tab', 'value', 'href'],
```
Comment (motorina0) [inline static/js/components.js]: - no need to pass the full charge data, the amount is sufficient
```suggestion
  props: ['charge-amount', 'tab', 'value', 'href'],
```

File: templates/satspay/display.html [hunk 17]
```diff
@@ -1,360 +1,234 @@
...
+              </div>
+              <satspay-show-qr
                 v-else
-                @click="payOnchain"
-                style="color: primary; width: 100%"
-                label="onchain⛓️"
-              >
-                <q-tooltip> pay onchain </q-tooltip>
-              </q-btn>
+                :charge="charge"
```
Comment (motorina0) [inline templates/satspay/display.html]:
```suggestion
                :charge-amount="charge.amount"
```

File: templates/satspay/display.html [hunk 18]
```diff
@@ -1,360 +1,234 @@
...
-                </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
               </div>
-
-              <a
-                class="text-secondary"
+              <satspay-show-qr
+                v-else
+                :charge="charge"
```
Comment (motorina0) [inline templates/satspay/display.html]:
```suggestion
                :charge-amount="charge.amount"
```

File: templates/satspay/display.html [hunk 19]
```diff
@@ -1,360 +1,234 @@
...
-                    @click="copyText(charge.onchainaddress)"
-                    >Copy address</q-btn
-                  >
-                </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
               </div>
+              <satspay-show-qr
+                v-else
+                :charge="charge"
```
Comment (motorina0) [inline templates/satspay/display.html]:
```suggestion
                :charge-amount="charge.amount"
```

File: static/js/components.js (hunk 16 above)
Comment (motorina0) [reply static/js/components.js]: - this component knows nothing about `tabs`, maybe the input should be named `type`
- `tab` is just an outer wrapper, in different contexts it can be something else (dialog, expandable-item, etc)

File: static/js/components.js (hunk 9 above)
Comment (talvasconcelos) [reply static/js/components.js]: i see your point. but these are specific components. removing the col size will make the QR smaller than "optimal"!

would it be ok if i pass the `class="col-sm-10 col-lg-8"` to the component? 

`<satspay-paid class="col-sm-10 col-lg-8">...</......>` ?

Comment (motorina0) [reply static/js/components.js]: why not just move the `<div col="...` at the parent level?

File: static/js/components.js (hunk 15 above)
Comment (talvasconcelos) [reply static/js/components.js]: yes it was wrong

File: static/js/components.js (hunk 9 above)
Comment (talvasconcelos) [reply static/js/components.js]: i did

File: static/js/components.js [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+          </span>
+          <span>{{charge.timeLeft}}</span>
+        </div>
+      </div>
+    </q-linear-progress>
+  </div>`,
+  created() {
+    if (!this.charge.timeLeft && !this.charge.paid) {
+      this.barText = 'Time elapsed'
+      return
```
Comment (motorina0) [inline static/js/components.js]:
```suggestion
```
- do not use return, later some more logic might be added to the end of `created()`

File: static/js/components.js [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    } else if (this.charge.paid) {
+      this.barText = 'Payment received'
+      return
```
Comment (motorina0) [inline static/js/components.js]:
```suggestion
```
- do not use return, later some more logic might be added to the end of `created()`

File: templates/satspay/display.html [hunk 22]
```diff
@@ -1,329 +1,210 @@
...
+          <q-item v-if="pendingFunds" dense>
+            <q-item-section>
+              <q-item-label>Amount pending:</q-item-label>
+              <!-- <q-item-label caption>Amount pending on mempool</q-item-label> -->
+            </q-item-section>
+
+            <q-item-section side>
               <q-badge color="gray">
-                <span v-text="pendingFunds" class="text-subtitle2"></span> sat
+                <span class="text-subtitle2">${pendingFunds} sats</span>
```
Comment (motorina0) [inline templates/satspay/display.html:54]: - is the `${pendingFunds}` expression evaluated inside an HTML?

File: templates/satspay/display.html [hunk 23]
```diff
@@ -1,329 +1,210 @@
...
-                  class="text-subtitle2"
-                ></span>
-                sat
+            </q-item-section>
+          </q-item>
+          <q-separator spaced></q-separator>
+          <q-item dense>
+            <q-item-section>
+              <q-item-label>Amount due:</q-item-label>
+              <!-- <q-item-label caption>Amount left to be paid</q-item-label> -->
```
Comment (motorina0) [inline templates/satspay/display.html]: - dead code
```suggestion
```

File: templates/satspay/display.html [hunk 24]
```diff
@@ -1,329 +1,210 @@
...
+            </q-item-section>
+
+            <q-item-section side>
+              <q-badge color="green">
+                <span class="text-subtitle2"
+                  >${charge.amount - charge.balance > 0 ? charge.amount -
```
Comment (motorina0) [inline templates/satspay/display.html:67]: - is `${...}` actually ok?

File: templates/satspay/display.html [hunk 25]
```diff
@@ -1,329 +1,210 @@
...
-            </div>
-          </div>
-          <div class="row items-center">
-            <div class="col-4 q-pr-lg">Total to pay:</div>
-            <div class="col-8 q-pr-lg">
+          </q-item-label>
+          <q-item dense>
+            <q-item-section>
+              <q-item-label>Total to pay:</q-item-label>
+              <!-- <q-item-label caption>Total amount to pay</q-item-label> -->
```
Comment (motorina0) [inline templates/satspay/display.html:23]:
```suggestion
```

File: templates/satspay/display.html [hunk 26]
```diff
@@ -1,329 +1,210 @@
...
-          <div class="row items-center q-mt-sm">
-            <div class="col-4 q-pr-lg">Amount paid:</div>
-            <div class="col-8 q-pr-lg">
+            </q-item-section>
+          </q-item>
+          <q-separator spaced></q-separator>
+          <q-item dense>
+            <q-item-section>
+              <q-item-label>Amount paid:</q-item-label>
+              <!-- <q-item-label caption>Amount already paid</q-item-label> -->
```
Comment (motorina0) [inline templates/satspay/display.html]:
```suggestion
```

File: templates/satspay/display.html [hunk 27]
```diff
@@ -1,329 +1,210 @@
...
-            <div class="col-8 q-pr-lg">
+                <span>&nbsp;sats</span>
+              </q-badge>
+            </q-item-section>
+          </q-item>
+          <q-separator spaced v-if="pendingFunds"></q-separator>
+          <q-item v-if="pendingFunds" dense>
+            <q-item-section>
+              <q-item-label>Amount pending:</q-item-label>
+              <!-- <q-item-label caption>Amount pending on mempool</q-item-label> -->
```
Comment (motorina0) [inline templates/satspay/display.html]:
```suggestion
```

File: templates/satspay/display.html [hunk 28]
```diff
@@ -367,6 +247,17 @@
       },
       amountDue() {
         return this.charge.amount - this.charge.balance
+      },
+      unifiedQR() {
+        const bitcoin = (this.charge.onchainaddress || '').toUpperCase()
+        let queryString = `bitcoin:${bitcoin}?amount=${
+          this.charge.amount / 1e8
```
Comment (motorina0) [inline templates/satspay/display.html:235]: - JS floating point `toString()` can be tricky:
image

File: templates/satspay/display.html [hunk 29]
```diff
@@ -1,329 +1,210 @@
...
+            label="Onchain"
+          ></q-tab>
+        </q-tabs>
+
+        <q-separator></q-separator>
+        <q-tab-panels v-model="tab" animated style="background: none">
+          <q-tab-panel name="uqr">
+            <div class="row justify-center q-mt-sm">
+              <div class="col-sm-10 col-md-8">
+                <div v-if="!charge.timeLeft && !charge.paid">
```
Comment (motorina0) [inline templates/satspay/display.html]: - this conditions repeats 3 times (once in each tab)
- it should be moved on the `<q-tab` level

```html
<div v-if="!charge.timeLeft && !charge.paid">
    <q-icon
      name="block"
      style="color: #ccc; font-size: 21.4em"
    ></q-icon>
</div>
<div v-else-if="charge.paid">
    <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
</div>
<q-tabs v-else ...>
   ...
</q-tabs>
```

**Comment by motorina0** [comment]
Scanning with Muun and Phoenix did not work. Maybe they don't support BIP21?

File: templates/satspay/display.html (hunk 24 above)
Comment (talvasconcelos) [reply templates/satspay/display.html:67]: it uses `delimiter` from vue, so we don't need the `{% raw %}` and can use JS

File: templates/satspay/display.html (hunk 28 above)
Comment (talvasconcelos) [reply templates/satspay/display.html:235]: better to explicitly do `this.charge.amount / 100000000` ?

**Comment by talvasconcelos** [comment]
> Scanning with Muun and Phoenix did not work. Maybe they don't support BIP21?

on Muun: "Go to Settings > Lightning Network > Receiving protocol and set it to"Unified" to enable this feature
Link to issue:"

Phoenix should work, according to this: https://bitcoinqr.dev/

I tried Phoenix, but don't have it on testnet! But it reads the QR

File: templates/satspay/display.html (hunk 28 above)
Comment (motorina0) [reply templates/satspay/display.html:235]: - it would give the same exponential result
- suggestion: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed

File: templates/satspay/display.html (hunk 22 above)
Comment (talvasconcelos) [reply templates/satspay/display.html:54]: It's javascript! It uses `delimiters` so we don't need to use `{% raw %}`

Comment (motorina0) [reply templates/satspay/display.html:54]: Its JS inside HTML. I didn't know this works.

File: templates/satspay/display.html (hunk 28 above)
Comment (motorina0) [reply templates/satspay/display.html:235]: This is not resolved, please have a look t [toFixed]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed) function

Comment (talvasconcelos) [reply templates/satspay/display.html:235]: ahh, i see the difference! pushed, `( this.charge.amount / 1e8).toFixed(8)`

**Review by motorina0 — APPROVED** [review]

**Review by dni — APPROVED** [review]
untested
